File: Data Structures & Algorithms/merge-intervals/submission-0.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, Comparator.comparingInt(a -> a[0]));
        List<int[]> result = new ArrayList<>();
        result.add(intervals[0]);

        for (int i = 1; i < intervals.length; i++) {
            if (intervals[i][0] <= result.get(result.size() - 1)[1]) {
                result.get(result.size() - 1)[0] = Math.min(intervals[i][0], result.get(result.size() - 1)[0]);
                result.get(result.size() - 1)[1] = Math.max(intervals[i][1], result.get(result.size() - 1)[1]);
            } else {{
                result.add(intervals[i]);
            }}
        }

        return result.toArray(new int[0][]);
    }
}

File: Data Structures & Algorithms/interval-list-intersections/submission-0.ts
class Solution {
    /**
     * @param {number[][]} firstList
     * @param {number[][]} secondList
     * @return {number[][]}
     */
    intervalIntersection(
        firstList: number[][],
        secondList: number[][],
    ): number[][] {
        const result: number[][] = [];
        let i = 0;
        let j = 0;
        

        while (i < firstList.length && j < secondList.length) {
            const start = Math.max(firstList[i][0], secondList[j][0]);
            const end = Math.min(firstList[i][1], secondList[j][1]);

            if (start <= end) {
                result.push([start, end]);
            }

            if (firstList[i][1] < secondList[j][1]) {
                i++;
            } else {
                j++;
            }
        }


        return result;
    }
}
